add visualize LSST galaxy-galaxy strong lensing simulation.

trivialTZ opened the pull request:
This proposal contains two files added. One is the simulation python file and one is the data yml file for pipeline used in the simulation.

trivialTZ committed:
yml file used in sim_lsst_galaxy_galaxy_lens_simulation

This file is a description for Skypy pipeline to generate the galaxy information observed by LSST. The file is used in sim_lsst_galaxy_galaxy_lens_simulation for galaxy-galaxy strong lensing visualize simulation.

trivialTZ committed:
galaxy-galaxy lens simulation for LSST

This file use Skypy pipeline and Lenstronomy to generate visualized galaxy-galaxy lensing pictures.

sibirrer commented on sim_pipeline/sim_lsst_galaxy_galaxy_lens_simulation.py:
```diff
@@ -0,0 +1,283 @@
+                            'e1': e1, 'e2': e2, 'center_x': center_x, 'center_y': center_y}]
+        return [kwargs_galaxy_g, kwargs_galaxy_r, kwargs_galaxy_i], distan, source['z'], stellar_mass
+
+def get_vd_from_stallermass(smass):
```
I would put this function in a different file that is tasked to link stellar mass and halo masses

sibirrer commented on sim_pipeline/sim_lsst_galaxy_galaxy_lens_simulation.py:
```diff
@@ -0,0 +1,283 @@
+
+def get_vd_from_stallermass(smass):
+    """
+    function for calculate the velocity from the staller mass for velocity dispersion
```
a bit a weird description. Also state that this is an empirical relation. You give the reference below but it deserves more explanations

sibirrer commented on sim_pipeline/sim_lsst_galaxy_galaxy_lens_simulation.py:
```diff
@@ -0,0 +1,283 @@
+        source = pipeline_cut[index]
+        return source
+
+    def get_params(self):
```
this function should have a more specific name indicating what is done here

sibirrer replied:
I would also split this function into 
1. where  some of the parameters are drawn with some randomness and
2. how these parameters are turned into lenstronomy compatible functions

sibirrer commented on sim_pipeline/sim_lsst_galaxy_galaxy_lens_simulation.py:
```diff
@@ -0,0 +1,283 @@
+        source= self.get_source()
+        stellar_mass = source['stellar_mass']
+        redz = source['z']
+        distan = [self.cosmo.angular_diameter_distance(redz).value]    # distance between galaxy and observer
```
why are there square brackets around the angular diameter distance?

sibirrer commented on sim_pipeline/sim_lsst_galaxy_galaxy_lens_simulation.py:
```diff
@@ -0,0 +1,283 @@
+            n_sersic = 1
+        else:
+            n_sersic = 4
+        center_x, center_y = np.random.uniform(-self.position_scatter, self.position_scatter), np.random.uniform(
```
this function I would either put somewhere else or have positional_scatter explicitly as an input argument.

sibirrer commented on sim_pipeline/sim_lsst_galaxy_galaxy_lens_simulation.py:
```diff
@@ -0,0 +1,283 @@
+from astropy.visualization import make_lupton_rgb
+from lenstronomy.SimulationAPI.ObservationConfig.LSST import LSST
+from lenstronomy.ImSim.image_model import ImageModel
+cosmo = FlatLambdaCDM(H0=67.66, Om0=0.30966)
```
I would split the routines and the specification of the cosmology, such that the user can chose them at input without changing the source code of this file

sibirrer commented on sim_pipeline/sim_lsst_galaxy_galaxy_lens_simulation.py:
```diff
@@ -0,0 +1,283 @@
+from lenstronomy.ImSim.image_model import ImageModel
+cosmo = FlatLambdaCDM(H0=67.66, Om0=0.30966)
+
+print('SkyPy version ==', skypy.__version__)
```
remove print statements

sibirrer commented on sim_pipeline/sim_lsst_galaxy_galaxy_lens_simulation.py:
```diff
@@ -0,0 +1,283 @@
+configblue = LenstronomyConfig(pipeline=skycon, mag_cut=22, z_min=0.8, z_max=2, galaxy_type='blue', position_scatter=0.5)
+configred = LenstronomyConfig(pipeline=skycon, mag_cut=20, z_min=0.1, z_max=0.4, galaxy_type='red', position_scatter=0.0)
+
+def process_image(ax, sim_g, sim_r, sim_i, cosmolo): #
```
this function mixes a lot of different things. I.e. source, lens light, mass model, lenstronomy configurations and plotting. I would suggest having a class or function that each deals with only one aspect, i.e. Source(), Lens(), then another class that casts the complete information into lenstronomy. And another class that does the plotting. This gives the best way to extend the code without breaking it.

sibirrer commented on sim_pipeline/sim_lsst_galaxy_galaxy_lens_simulation.py:
```diff
@@ -0,0 +1,283 @@
+        v_sigma_theta_Evalues['list1'].append(vsigma)
+        v_sigma_theta_Evalues['list2'].append(theE)
+
+f.tight_layout()
```
I would put these plotting routines in a different python file with is only dealing with plotting

sibirrer reviewed (changes requested):
Thank you very much @trivialTZ !
As you can see I suggest quite some changes in the structure. Try to have each class or definition do only one specific task. Also try to split the source and lens description, as well as what is a complete lens and source, and how it is being cast into lenstronomy (imaging someone else wants to use a different lensing code but wants to use the other parts, or change the source model but wants to keep all other parts).

Then about testing:
- make the same structure of files in the 'tests' folder with 'test_<filename>' and then find at least one example of a test that runs and makes the test pass.

sibirrer commented on data/SkyPy/lsst-like.yml:
```diff
@@ -0,0 +1,100 @@
+mag_lim: 35
```
add a readme.md file in the same folder to describe this file.

sibirrer replied:
This is an excellent readme. Thank you very much!

sibirrer commented on sim_pipeline/sim_lsst_galaxy_galaxy_lens_simulation.py:
```diff
@@ -0,0 +1,283 @@
+        size_arcsec = source['angular_size'] / constants.arcsec   # convert radian to arc seconds
+        mag_g, mag_r, mag_i = source['mag_g'], source['mag_r'], source['mag_i']
+        mag_apparent = source['M']
+        e = source['ellipticity']
```
is this axis ratio or eccentricity? Check skypy. For conventions, see: https://github.com/lenstronomy/lenstronomy-tutorials/blob/main/Notebooks/GettingStarted/units_coordinates_parameters_definitions.ipynb

trivialTZ committed:
Create readme.md

trivialTZ committed:
Improve galaxy-galaxy-lens; Add a Jupyter notebook for how to use it; Add get_vd_from_stallermass function;

For g_g_lensing:
1. The get_galaxyparas_from_pipeline class is created for obtaining galaxy parameters, including galaxy distances, redshifts, and band information for g, r, and i bands, from the Lenstronomy Pipeline.
GGLensingSystem_kwargs, is added to execute the galaxies-galaxies lensing system information, including source galaxies information, lensing galaxies information, lensing-system information like the Einstein radius, and the lensing picture information.
2. The GGlensing_image class is introduced to simulate gravitational lensing images in g, r, and i bands using the provided configurations for source (blue) and lens (red) galaxies. This class also supports generating images with noise using the get_irg_noise() method.
3. The GGlensing_images_plot class is added for creating and displaying simulated gravitational lensing images using the provided configurations for source (blue) and lens (red) galaxies. This class supports plotting single images with or without noise using the plot_one_image() method. Using plot_images method to draw multiple images.
//////////////////////////////////////////////////////////////////////
For gg_lensing_tutorial:
1. Create a Skypipeline class that initializes and executes a SkyPy pipeline using a given configuration file. The run_pipeline method executes the pipeline and returns its state.
2.Define the file paths and directories needed to access the SkyPy configuration file and data. Read the SkyPy configuration file and execute the pipeline using the Skypipeline class, storing the pipeline's state in the skycon variable.
Use the get_galaxyparas_from_pipeline function to obtain galaxy parameters for blue and red galaxies. This includes applying magnitude and redshift limits, specifying galaxy types, and adding position scatter.
3.Create an instance of the GGlensing_images_plot class with the galaxy parameters. Call the plot_images method on the ggl_images_plot instance to generate and display the galaxy images with lensing effects. This includes specifying the layout of the images (6 horizontal and 3 vertical) and adding noise to the images.
////////////////////////////////////////////////////////////////////////
For get_vd_from_stallermass:
This function allows users to easily calculate the velocity dispersion of a galaxy based on its stellar mass, using a well-established empirical relationship from the literature.

trivialTZ commented:
Improve galaxy-galaxy-lens; Add a Jupyter notebook for how to use it; Add get_vd_from_stallermass function;
///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
For g_g_lensing:
1. The get_galaxyparas_from_pipeline class is created for obtaining galaxy parameters, including galaxy distances, redshifts, and band information for g, r, and i bands, from the Lenstronomy Pipeline.
GGLensingSystem_kwargs, is added to execute the galaxies-galaxies lensing system information, including source galaxies information, lensing galaxies information, lensing-system information like the Einstein radius, and the lensing picture information.
2. The GGlensing_image class is introduced to simulate gravitational lensing images in g, r, and i bands using the provided configurations for source (blue) and lens (red) galaxies. This class also supports generating images with noise using the get_irg_noise() method.
3. The GGlensing_images_plot class is added for creating and displaying simulated gravitational lensing images using the provided configurations for source (blue) and lens (red) galaxies. This class supports plotting single images with or without noise using the plot_one_image() method. Using plot_images method to draw multiple images.
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
For gg_lensing_tutorial:
1. Create a Skypipeline class that initializes and executes a SkyPy pipeline using a given configuration file. The run_pipeline method executes the pipeline and returns its state.
2.Define the file paths and directories needed to access the SkyPy configuration file and data. Read the SkyPy configuration file and execute the pipeline using the Skypipeline class, storing the pipeline's state in the skycon variable.
Use the get_galaxyparas_from_pipeline function to obtain galaxy parameters for blue and red galaxies. This includes applying magnitude and redshift limits, specifying galaxy types, and adding position scatter.
3.Create an instance of the GGlensing_images_plot class with the galaxy parameters. Call the plot_images method on the ggl_images_plot instance to generate and display the galaxy images with lensing effects. This includes specifying the layout of the images (6 horizontal and 3 vertical) and adding noise to the images.
///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
For get_vd_from_stallermass:
This function allows users to easily calculate the velocity dispersion of a galaxy based on its stellar mass, using a well-established empirical relationship from the literature.

sibirrer reviewed (approved):
Thank you @trivialTZ !
I have quite some comments about design and also don't see any test functions.
I think I merge it and take it from there for the next 1-2 days to re-factor the design a bit
